### PR TITLE
Managed TLS (lets encrypt)

### PR DESCRIPTION
**Pull request by carbon198**

### CODE REVIEW

File: lib/fastlyctl/commands/tls/managed.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
+module FastlyCTL
+  class TLSManagedSubCmd < SubCommandBase
+    SubcommandPrefix = "tls managed"
```
Comment (denen99) [inline lib/fastlyctl/commands/tls/managed.rb:3]: Stupid nitpick, but is our new TLS self-service still called "managed"?  We have Free, Platform and Concierge coupled with Fastly managed certs, BYO certs, etc.  The whole thing is a bit confusing still. Point being, i want to make sure you leave yourself enough flexibility here for all the diff TLS offerings we have.  If managed is the right choice, all good, just wanted to sanity check.

Comment (carbon198) [reply lib/fastlyctl/commands/tls/managed.rb:3]: It's referencing this: https://docs.fastly.com/api/tls-subscriptions

The title on that page is `Fastly-managed TLS subscriptions`. If we are calling it something else I can easily swap the name to whatever it should be.

Comment (carbon198) [reply lib/fastlyctl/commands/tls/managed.rb:3]: I was planning on having different subcommands for different categories. For instance, `platform` for Platform TLS, and `managed` for managed certificate subscriptions. Additionally, the other API endpoints can just be called what they are: `private` for private keys, `domain` for domains, `cert` for certificates, and `activation` for activations.

Comment (carbon198) [reply lib/fastlyctl/commands/tls/managed.rb:3]: Would you rather I called them `subscriptions`? "Subscription" is the noun while "managed" is the modifier, so gramatically it probably does make more sense to call them subscriptions.

**Review by denen99 — APPROVED** [review]
Looks good.

File: lib/fastlyctl/commands/tls.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  module TLSUtils
+    def self.get_tls_configs
+      data = FastlyCTL::Fetcher.api_request(:get,"/tls/configurations",{use_vnd:true})["data"]
+      if data.length == 0
+        thor = Thor::Shell::Basic.new
+        thor.say "No TLS Configurations found. You may need to upgrade to a paid account if you are using a free account."
+        thor.say "If you need assistance, please contact support@fastly.com."
+        if (thor.yes?("Would you like to open the TLS configuration page in the Fastly app?"))
+          FastlyCTL::Utils.open_app_path("/network/domains")
```
Comment (drwilco) [inline lib/fastlyctl/commands/tls.rb:24]: Isn't this `/network/tls/domains`?

File: lib/fastlyctl/commands/tls/managed.rb [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    DomainRegex = /(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]/
+
+    desc "create <domain>", "Create a Fastly Managed TLS Subscription for [domain]. A Certificate will be requested from lets-encrypt once you satisfy one of the challenges. You can learn more about the challenge types here: https://letsencrypt.org/docs/challenge-types/ and Fastly's API documentation here: https://docs.fastly.com/api/tls-subscriptions."
+    def create(domain)
+      abort "Must specify valid domain name" unless domain =~ DomainRegex
+
+      tls_configs = FastlyCTL::TLSUtils.get_tls_configs
+      tls_config = FastlyCTL::TLSUtils.select_tls_config(tls_configs)
```
Comment (drwilco) [inline lib/fastlyctl/commands/tls/managed.rb:11]: Allow this to be done by cli parameter?

File: lib/fastlyctl/fetcher.rb [hunk 4]
```diff
@@ -54,25 +64,41 @@ def self.api_request(method, path, options={})
...
         end
 
         error += " Method: #{method.to_s.upcase}, Path: #{path}\n"
-        error += "Message from API: #{response.response_body}"
+
+        if (options[:use_vnd]) 
+          begin
+            error_resp = JSON.parse(response.response_body)
+          rescue JSON::ParserError
+            error_resp = {"errors" => [{"title" => "Error parsing response JSON","details" => "No further information available. Please file a github issue at https://github.com/fastly/fastlyctl"}]}
```
Comment (drwilco) [inline lib/fastlyctl/fetcher.rb:87]: Does this output the JSON in question?

**Review by drwilco — CHANGES_REQUESTED** [review]
I'm not great at Ruby, so not the most thorough review

File: lib/fastlyctl/commands/tls.rb (hunk 2 above)
Comment (carbon198) [reply lib/fastlyctl/commands/tls.rb:24]: Not to my knowledge.

File: lib/fastlyctl/commands/tls/managed.rb (hunk 3 above)
Comment (carbon198) [reply lib/fastlyctl/commands/tls/managed.rb:11]: Ah, good call. I will add that during the next pass.

File: lib/fastlyctl/fetcher.rb (hunk 4 above)
Comment (carbon198) [reply lib/fastlyctl/fetcher.rb:87]: In this particular code path there is no JSON. This line is for a JSON parse error.

**Pull request merged**